### PR TITLE
3.6.3: IMPORTANT! Preventing undesired access to guests/registered

### PR DESCRIPTION
**Pull request by infograf768**
Pull Request for Issue https://github.com/joomla/joomla-cms/pull/12353#issuecomment-254013691

Adding back checks to prevent unwanted access to details from non-authors.

### CODE REVIEW

File: components/com_content/content.php [hunk 1]
```diff
@@ -12,6 +12,37 @@
...
+elseif ($input->get('view') === 'articles' && $input->get('layout') === 'modal')
+{
+	if (!$user->authorise('core.create', 'com_content'))
+	{
+		JFactory::getApplication()->enqueueMessage(JText::_('JERROR_ALERTNOAUTHOR'), 'warning');
+
+		return;
+	}
+}
+elseif ($input->get('view') === 'module' && $input->get('layout') === 'modal')
```
Comment (mbabker) [inline components/com_content/content.php]: Umm, this file shouldn't ever handle a request for modules?

Comment (infograf768) [inline components/com_content/content.php]: we have a xd module since some times ago.
By default, Joomla gives Create rights to authors.

**Comment by Hackwar** [comment]
can we please add the tag for release blocker to this one?

File: components/com_content/content.php (hunk 1 above)
Comment (zero-24) [inline components/com_content/content.php]: but this file (components/com_content/content.php) should not handle that module button?

Comment (mbabker) [inline components/com_content/content.php]: Right, but the plugin is requesting com_modules.  So the com_content "front controller" shouldn't be processing this check unless something's massively hacked...

**Comment by infograf768** [comment]
At @mbabker request, taking off check for module access

**Comment by mbabker** [comment]
The check is right, it should just be in `components/com_modules/modules.php`.

**Comment by infograf768** [comment]
On it

**Comment by infograf768** [comment]
Done. Please test asap and mark results on issues.joomla.org

**Review by mbabker — APPROVED** [review]

**Comment by zero-24** [comment]
I have tested this item :red_circle: unsuccessfully on 8918b5926736719eae062d090847ee217355e361<br /><br />The sec issue still exists.

### Get the list of modules without login

go to the login screen
see the html code for the token
use this URL: http://example.org/index.php?option=com_modules&view=modules&layout=modal&tmpl=component&editor=jform_articletext&<token>=1<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/12435">issues.joomla.org/tracker/joomla-cms/12435</a>.</sub>

File: components/com_modules/modules.php [hunk 2]
```diff
@@ -12,10 +12,21 @@
 // Load the required admin language files
 $lang   = JFactory::getLanguage();
 $app    = JFactory::getApplication();
+$user   = JFactory::getUser();
 $config = array();
 $lang->load('joomla', JPATH_ADMINISTRATOR);
 $lang->load('com_modules', JPATH_ADMINISTRATOR);
 
+if ($app->input->get('view') === 'module' && $app->input->get('layout') === 'modal')
```
Comment (zero-24) [inline components/com_modules/modules.php]: it looks like this needs to be modules?

**Comment by dgt41** [comment]
@infograf768 check https://github.com/infograf768/joomla-cms/pull/44

File: components/com_content/content.php [hunk 3]
```diff
@@ -12,6 +12,28 @@
...
 JLoader::register('ContentHelperQuery', JPATH_SITE . '/components/com_content/helpers/query.php');
 
+$input = JFactory::getApplication()->input;
+$user  = JFactory::getUser();
+
+if ($input->get('view') === 'article' && $input->get('layout') === 'pagebreak')
+{
+	if (!$user->authorise('core.create', 'com_content'))
+	{
+		JFactory::getApplication()->enqueueMessage(JText::_('JERROR_ALERTNOAUTHOR'), 'warning');
```
Comment (andrepereiradasilva) [inline components/com_content/content.php]: shouldn't we use the new exceptions in this messages?
```php
throw new JAccessExceptionNotallowed(JText::_('JERROR_ALERTNOAUTHOR'), 403);
```

File: components/com_content/content.php [hunk 4]
```diff
@@ -12,6 +12,28 @@
 JLoader::register('ContentHelperRoute', JPATH_SITE . '/components/com_content/helpers/route.php');
 JLoader::register('ContentHelperQuery', JPATH_SITE . '/components/com_content/helpers/query.php');
 
+$input = JFactory::getApplication()->input;
+$user  = JFactory::getUser();
+
+if ($input->get('view') === 'article' && $input->get('layout') === 'pagebreak')
+{
+	if (!$user->authorise('core.create', 'com_content'))
```
Comment (andrepereiradasilva) [inline components/com_content/content.php]: what if:
- the user is not allowed to `core.create`, but is allowed to `core.edit`
- the user is not allowed to `core.create`, but is allowed to `core.edit.own` and this content was created by him
- etc

will this work?

Comment (mbabker) [inline components/com_content/content.php]: These ACL checks need a lot of work in general.  Unless we're going to hold up 3.6.3 to get them 100% right (they haven't ever been), I'd suggest going with this as is.

Comment (andrepereiradasilva) [inline components/com_content/content.php]: yes, if i'm not wrong you would need, at leaset, something like this 
```php
$canCreate  = $user->authorise('core.create',     'com_content.category.' . $item->catid);
$canEdit    = $user->authorise('core.edit',       'com_content.article.' . $item->id);
$canEditOwn = $user->authorise('core.edit.own',   'com_content.article.' . $item->id) && $item->created_by == $user->id;

if (($canCreate && $item->id === 0) || ($canEdit && $item->id !== 0) || ($canEditOwn && $item->id !== 0))
{
 // ...
}
```
in other words you need the user id, the item id, the item created by id and the item category id to perform those checks

Comment (mbabker) [inline components/com_content/content.php]: Pretty much.  Except you won't have all of that when you're creating new items, so you have to account for that too.

Comment (andrepereiradasilva) [inline components/com_content/content.php]: yes, of course, i think for new items you just need to check the core.create in the category the article is

Comment (Hackwar) [inline components/com_content/content.php]: You don't have an item (=article) at that point yet. So far, you only opened the edit view and next clicked on the button. There is neither an ID nor a category.

File: components/com_modules/modules.php (hunk 2 above)
Comment (infograf768) [inline components/com_modules/modules.php]: Yes, correcting now

**Comment by infograf768** [comment]
@mbabker 
>  It's totally possible to have a group with create permissions without edit, so you're more likely breaking something than fixing it (compared to 3.6.2 which checks core.edit).

The aim of all this was to let authors use these xtds. They do not have edit permissions which prevented them from using the xtds in front-end.
One could indeed have edit permissions without create, but this is very unlikely.

Therefore the main difference with 3.6.2 is that authors now can see the lists, use pagebreak and that the modules modal is access protected.

I have now corrected the PR for typo (module(s) and too much vars.
I propose to go with it and fine grain ACL in 3.7.

**Comment by zero-24** [comment]
I have tested this item :white_check_mark: successfully on 71b5544ead96782600f26a5d118e263480582724<br /><br />Works here. Just a cosmetic change that the error message should be a error and not a warning. But other things works great!<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/12435">issues.joomla.org/tracker/joomla-cms/12435</a>.</sub>

**Comment by infograf768** [comment]
Warning is what we got before 
https://github.com/joomla/joomla-cms/pull/12353/files#diff-35390bcd97e9f612d6fc06ea874aa22aL22

Let's fine tune all this in 3.7 :)

**Comment by zero-24** [comment]
:+1:

**Comment by mbabker** [comment]
> The aim of all this was to let authors use these xtds. They do not have edit permissions which prevented them from using the xtds in front-end.

If every site uses the default ACL system then thinking in that structure is valid.  Except not every site uses the default ACL structure.  Many of the sites I manage have most of the default groups deleted and custom groups added.  In general we have to think beyond the ACL groups that are shipped as part of our dataset and look at the actual permissions when we are doing things like this.

**Pull request merged**